test/system/src/test/java/io/pravega/test/system/BookieFailoverTest.java

```diff
@@ -144,40 +145,50 @@ public void tearDown() {
     }
 
     @Test
-    public void bookieFailoverTest() throws ExecutionException, InterruptedException {
+    public void bookieFailoverTest() throws Exception {
         createWriters(clientFactory, NUM_WRITERS, SCOPE, STREAM);
         createReaders(clientFactory, readerGroupName, SCOPE, readerGroupManager, STREAM, NUM_READERS);
 
         // Give some time to create readers before forcing a bookie failover.
-        Exceptions.handleInterrupted(() -> Thread.sleep(BOOKIE_FAILOVER_WAIT_MILLIS));
+        AssertExtensions.assertEventuallyEquals("Writers and/or readers not progressing.", true,
+            () -> testState.getEventWrittenCount() > 100 && testState.getEventReadCount() > 100, 1000, IO_PROGRESS_CHECK_TIMEOUT);
 
         // Scale down bookie.
         Futures.getAndHandleExceptions(bookkeeperService.scaleService(2), ExecutionException::new);
 
-        log.info("Sleeping for {} seconds.", BOOKIE_FAILOVER_WAIT_MILLIS / 1000);
-        Exceptions.handleInterrupted(() -> Thread.sleep(BOOKIE_FAILOVER_WAIT_MILLIS));
-
-        long writeCountBeforeSleep  = testState.getEventWrittenCount();
-        long readCountBeforeSleep  = testState.getEventReadCount();
-        log.info("Write count is {} and read count is {} after {} seconds sleep after bookie failover.", writeCountBeforeSleep,
-                readCountBeforeSleep, BOOKIE_FAILOVER_WAIT_MILLIS / 1000);
+        log.info("Waiting until writers and readers cannot make further progress.");
+        long writeCountDiff, readCountDiff;
+        do {
+            writeCountDiff = testState.getEventWrittenCount();
+            readCountDiff = testState.getEventReadCount();
+            Exceptions.handleInterrupted(() -> Thread.sleep(2000));
+            writeCountDiff = testState.getEventWrittenCount() - writeCountDiff;
+            readCountDiff = testState.getEventReadCount() - readCountDiff;
+            log.info("Written ({}) and read ({}) events after downscale within 2 seconds.", writeCountDiff, readCountDiff);
+        } while (writeCountDiff + readCountDiff > 0);
```

Comment:
It will only ensure either read or write is happening, not both. Is that enough?

Reply:
No, this is going to leave the loop only if neither writes or reads are happening. Here we want to ensure that readers and writers cannot make further progress because of the downscale of Bookkeeper. Once this happens, we can assume that the system cannot execute IO and the effect of the Bookie failover is ongoing. The next assertion after the sleep checks that this is actually true for `BOOKIE_FAILOVER_WAIT_MILLIS` milliseconds.